### PR TITLE
Add UI for disabling widget replacement for particular widgets

### LINKED ISSUE #2049: Ability to disable widget replacement for some widgets
Ability to block some icons completely (for example, someone would not want to see Facebook like buttons if he doesn't use Facebook) and show some icons as Privacy Badger ones.

### PR DESCRIPTION
**Pull request by JacobHearst**
Fixes #2049 by adding the option to completely block some social
widgets instead of replacing them

### CODE REVIEW

**Comment by JacobHearst** [comment]
Hi, thanks for the quick review! I'm not sure that I'm following your first two bullets.

- Widget replacement is still opt-out, are you saying that all widgets should be blocked by default?
- From what I understand of the widget replacement code, if a widget isn't passed to the replacement content script, it will just leave the omitted widget alone which means that widgets which are supposed to be blocked will still appear and will be unchanged.

This is my first contribution to this project and it's very possible that I'm missing or misunderstanding something so please correct me if I've misstated something.

**Comment by ghostwords** [comment]
> it will just leave the omitted widget alone which means that widgets which are supposed to be blocked will still appear and will be unchanged

Widget replacement happens [only for blocked/"cookieblocked" domains](https://github.com/EFForg/privacybadger/blob/6e9650def2aad2b29d861b17816068a7049ab1ba/src/js/webrequest.js#L530). In the future, it will happen for blocked domains only.

>Widget replacement is still opt-out

I guess I'm confused by the "Widgets to hide" label. We should probably reword it/rework the UI to make it more clear that this is a way to exclude widgets from being replaced. We should avoid introducing new language/concepts such as "hide" if we can help it. Privacy Badger blocks domains. Some of these domains are linked to potentially useful widgets. Some of these widgets have replacements in Privacy Badger. Privacy Badger will replace them by default, but you can turn off replacement entirely, or turn it off for particular widgets (this PR).

**Comment by ghostwords** [comment]
By the way, if it gets too cramped on the General Settings tab, we could move widgets to their own tab.

If we do go with a new tab, we shouldn't call it Social Widgets as not all widgets at this point are social buttons.

**Comment by JacobHearst** [comment]
I've moved the widgets to their own tab under the settings page and hopefully cleared up some of the language but I'm still confused by this:

> Do we need to modify anything in the widget replacement content script? We can just check for individual widget preferences at the time of sending data to the content script, and not send data for widgets that have been deselected

If I mock not sending data for certain widgets by modifying the `initialize` function in the socialwidgets content script to this:

    function initialize() {
      // Get tracker info and check for initial blocks (that happened
      // before content script was attached)
      getTrackerData(function (trackers, trackerButtonsToReplace, trackerButtonsToBlock) {
        trackerInfo = trackers;
        delete trackerButtonsToReplace["Facebook Like"];
        replaceInitialTrackerButtonsHelper(trackerButtonsToReplace, trackerButtonsToBlock);
      });
    ...
    }

And then navigate to a page with a Facebook Like button on it, the like button is present in its original, unmodified form. Simply not sending back data for a widget doesn't mean that the content script will remove it from the page, it'll just ignore it. Am I just misunderstanding your comment?

**Comment by ghostwords** [comment]
Yes ... although it's confusing to me to refer to widgets we do not want replaced "blocked". I suggest renaming `blockedSocialWidgets` to something like `widgetReplacementExceptions`. (No "social" because not all widgets are social buttons.)

**Comment by JacobHearst** [comment]
> Yes

Fantastic, sorry for all the confusion on my part over the language and the content script. I've updated all `blockedSocialWidgets` variable/property names to `widgetReplacementExceptions`, are there any other changes you'd like to see?

Also, apologies for being so slow to respond on this PR and get updates in, I've been incredibly busy with school as of late and haven't had a lot of time or brainpower for programming outside of school

File: src/_locales/en_US/messages.json [hunk 1]
```diff
@@ -594,5 +598,13 @@
...
         "message": "Privacy Badger doesn't work on special pages like this one. Try browsing somewhere else.",
         "description": "Shown in the popup for special browser pages such as the New Tab page and 'about:' pages."
+    },
+    "widget_control_tab": {
+        "message": "Widgets",
+        "description": "Options page tab heading"
+    },
+    "widget_blocking_desc": {
+        "message": "Privacy Badger automatically replaces widgets with trackers embedded in them. If there are widgets that you'd like to be removed rather than replaced, you can manage that here.",
+        "description": "Description of Privacy Badger widget blocking and replacement"
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: Should instead describe where this text can be found (on the options page under the Widget Replacement tab) by translators.

File: src/_locales/en_US/messages.json [hunk 2]
```diff
@@ -594,5 +598,13 @@
     "popup_special_page_paragraph": {
         "message": "Privacy Badger doesn't work on special pages like this one. Try browsing somewhere else.",
         "description": "Shown in the popup for special browser pages such as the New Tab page and 'about:' pages."
+    },
+    "widget_control_tab": {
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: Naming nit: `options_widget_replacement_tab`

File: src/_locales/en_US/messages.json [hunk 3]
```diff
@@ -131,8 +131,12 @@
         "message": "Import",
         "description": ""
     },
+    "options_hide_social_widgets": {
+        "message": "Widgets to remove:",
+        "description": "Multiple selection box on the general settings page"
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: Under the Widget Replacement tab now.

File: src/_locales/en_US/messages.json [hunk 4]
```diff
@@ -594,5 +598,13 @@
     "popup_special_page_paragraph": {
         "message": "Privacy Badger doesn't work on special pages like this one. Try browsing somewhere else.",
         "description": "Shown in the popup for special browser pages such as the New Tab page and 'about:' pages."
+    },
+    "widget_control_tab": {
+        "message": "Widgets",
+        "description": "Options page tab heading"
+    },
+    "widget_blocking_desc": {
+        "message": "Privacy Badger automatically replaces widgets with trackers embedded in them. If there are widgets that you'd like to be removed rather than replaced, you can manage that here.",
```
[the remark this answers is not in the document]
Comment (JacobHearst) [reply src/_locales/en_US/messages.json]: "If there are buttons you don't want replaced, you can update your preferences here."?

**Comment by JacobHearst** [comment]
I believe I've addressed everything

File: src/_locales/en_US/messages.json (hunk 1 above)
Comment (ghostwords) [reply src/_locales/en_US/messages.json]: Why don't we say something like, Introduction to the Widget Replacement tab on the options page.

File: src/_locales/en_US/messages.json [hunk 5]
```diff
@@ -594,5 +598,13 @@
     "popup_special_page_paragraph": {
         "message": "Privacy Badger doesn't work on special pages like this one. Try browsing somewhere else.",
         "description": "Shown in the popup for special browser pages such as the New Tab page and 'about:' pages."
+    },
+    "options_widget_replacement_tab": {
+        "message": "Widget Replacement",
+        "description": "Options page tab heading"
+    },
+    "options_widget_replacement_desc": {
+        "message": "Privacy Badger can replace social buttons and other potentially useful widgets with click-to-activate placeholders. If there are buttons you don't want replaced, you can update your preferences here.",
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: Let's just remove the second sentence.

File: src/_locales/en_US/messages.json [hunk 6]
```diff
@@ -131,8 +131,12 @@
         "message": "Import",
         "description": ""
     },
+    "options_hide_social_widgets": {
+        "message": "Don't replace these widgets:",
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: "the following widgets" instead?

File: src/skin/options-layout.css [hunk 7]
```diff
@@ -153,6 +153,16 @@ header, #tabs {
...
 }
 
+#hide-widgets-row {
+    width: 100%;
+    margin-left: 25px;
+    margin-top: 5px;
+}
+
+#hide-widgets-select {
+  width: 500px;
```
Comment (ghostwords) [inline src/skin/options-layout.css]: I think this is too large for mobile (Firefox for Android)/small displays. Previously: #2195.

**Comment by JacobHearst** [comment]
> Where did this Select2 build come from?

I think I grabbed the "full" version, they just released 4.0.11 a few days ago so I've just updated to that version instead which I grabbed from [here](https://github.com/select2/select2/releases/tag/4.0.11)

File: src/skin/options-layout.css (hunk 7 above)
Comment (JacobHearst) [reply src/skin/options-layout.css]: This is what it looks like after updating the width to 300 px:

image

File: src/_locales/en_US/messages.json [hunk 8]
```diff
@@ -594,5 +598,13 @@
     "popup_special_page_paragraph": {
         "message": "Privacy Badger doesn't work on special pages like this one. Try browsing somewhere else.",
         "description": "Shown in the popup for special browser pages such as the New Tab page and 'about:' pages."
+    },
+    "options_widget_replacement_tab": {
+        "message": "Widget Replacement",
+        "description": "Options page tab heading"
+    },
+    "options_widget_replacement_desc": {
+        "message": "Privacy Badger can replace social buttons and other potentially useful widgets with click-to-activate placeholders.",
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: What do you think about the following text?

>When blocking social buttons and other potentially useful (video, audio, comments) widgets, Privacy Badger can replace them with click-to-activate placeholders.

- Tries to explain "widgets"
- Tries to explain that this isn't what blocks the widgets; the widgets are blocked, and then the replacement may or may not happen.

File: src/_locales/en_US/messages.json [hunk 9]
```diff
@@ -131,8 +131,12 @@
...
         "description": ""
     },
+    "options_hide_social_widgets": {
+        "message": "Don't replace the following widgets:",
+        "description": "Multiple selection box on the widget replacement tab"
+    },
     "options_social_widgets_checkbox": {
-        "message": "Replace social widgets",
+        "message": "Enable widget replacement",
         "description": "Checkbox label on the general settings page"
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: It's now on the Widget Replacement tab on the options page.

File: src/_locales/en_US/messages.json (hunk 8 above)
Comment (JacobHearst) [reply src/_locales/en_US/messages.json]: I like it

File: src/_locales/en_US/messages.json (hunk 9 above)
Comment (JacobHearst) [reply src/_locales/en_US/messages.json]: Ah, yes

**Pull request merged**